zipline/algorithm.py

```diff
@@ -440,17 +440,28 @@ def _create_generator(self, sim_params):
             self.initialize(*self.initialize_args, **self.initialize_kwargs)
             self.initialized = True
 
-        self.trading_client = self._create_trading_client(sim_params)
-
-        return self.trading_client.transform()
-
-    def _create_trading_client(self, sim_params):
-        return AlgorithmSimulator(
+        self.trading_client = AlgorithmSimulator(
             self,
             sim_params,
             self.data_portal,
             self._create_clock(),
-            self._create_benchmark_source()
+            self._create_benchmark_source(),
+            universe_func=self._calculate_universe
+        )
+
+        return self.trading_client.transform()
+
+    def _calculate_universe(self):
+        # this exists to provide backwards compatibility for older,
+        # deprecated APIs, particularly around the iterability of
+        # BarData (ie, 'for sid in data).
+
+        # our universe is the intersection of:
+        # 1) all assets in our positions
+        # 2) all assets in open orders
+        return set(
+            list(self.perf_tracker.position_tracker.positions.keys()) +
```

Comment:
If you're treating the keys of a dictionary as a set-like object, you probably want `viewkeys`: https://docs.python.org/2/library/stdtypes.html#dictionary-view-objects

`self.perf_tracker.position_tracker.positions.viewkeys() | self.blotter.open_orders.viewkeys()` is equivalent to this, but without the repeated traversals.